### PR TITLE
Record every training loss. Validate often. Use batch-based timesteps.

### PR DESCRIPTION
**Pull request by mrcslws**
This change:

- Adds mixins that log every training loss to the result dict
  - There is very little performance overhead; they are copied from the GPU at the end of the epoch.
  - Similarly, I log the learning rate, momentum, and regularization info.
- Changes the `ImagenetExperiment` to support multiple validations per epoch.
- Adds logic for converting an epoch result dict to a time series, and adds support to our `WandbLogger`

With this change I am making it more common for us to put large data structures in the result dict, so I added logic to ensure those large data structures don't get printed to the console.

### CODE REVIEW

File: nupic/research/frameworks/pytorch/imagenet/mixins/regularize_loss.py [hunk 1]
```diff
@@ -81,22 +84,58 @@ def pre_batch(self, model, batch_idx):
...
         self.reg_scalar = self._compute_reg_scalar(pct)
+        self.reg_scalar_history.append(self.reg_scalar)
 
     def complexity_loss(self, model):
         c_loss = super().complexity_loss(model)
         assert model is self.model
         reg = torch.stack([module.regularization()
                            for module in self._regularized_modules]).sum()
+        if self.model.training:
+            self.model_complexity_history.append(reg.detach().clone())
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/imagenet/mixins/regularize_loss.py:95]: The same GPU memory issue as complexity_loss
